### PR TITLE
adapter.http: AAS `SpecificAssetId` filtering

### PR DESCRIPTION
**Pull request by hadijannat**
The procedure is as follows:

- It first extracts the specific Asset IDs from the request,
- then transforms the base64url-encoded specificAssetIds into JSON and decodes the JSON via the HTTPApiDecoder to SpecificAssetId instances.
- Finally, it uses the SpecificAssetId instances to filter the AAS contained in the object store and only returns AAS that match all SpecificAssetIds.

First, it ensures that we have an appropriate method in HTTPApiDecoder to handle decoding base64url-encoded strings into JSON and then into SpecificAssetId instances.

```
import base64
import json

class HTTPApiDecoder:
    # Existing class content...

    @staticmethod
    def decode_specific_asset_ids(encoded_ids: str):
        # Decoding from base64url and transforming to JSON
        decoded_json = base64.urlsafe_b64decode(encoded_ids + '===').decode('utf-8')
        # Assuming the JSON represents a list of SpecificAssetId objects
        specific_asset_ids = json.loads(decoded_json)
        # Transform the list of dictionaries into SpecificAssetId instances
        return [model.SpecificAssetId(**id_dict) for id_dict in specific_asset_ids]
```
Finally, the modified get_aas_all function to use this new method for filtering based on SpecificAssetIds is like below:

```
def get_aas_all(self, request: Request, url_args: Dict, **_kwargs) -> Response:
    response_t = get_response_type(request)
    aas: Iterator[model.AssetAdministrationShell] = self._get_all_obj_of_type(model.AssetAdministrationShell)

    # Filter by 'idShort' if the parameter is provided in the request
    id_short = request.args.get("idShort")
    if id_short is not None:
        aas = filter(lambda shell: shell.id_short == id_short, aas)
    
    # Filtering by base64url encoded SpecificAssetIds if provided
    asset_ids = request.args.get("assetIds")
    if asset_ids is not None:
        # Use the new method to decode and instantiate SpecificAssetIds
        spec_asset_ids = HTTPApiDecoder.decode_specific_asset_ids(asset_ids)
        # Filter the AAS based on these SpecificAssetIds
        aas = filter(lambda shell: all(spec_id in [id_.identifier for id_ in shell.specific_asset_ids] for spec_id in spec_asset_ids), aas)
    
    aas_list = list(aas)
    return response_t(aas_list)
```

### CODE REVIEW

File: basyx/aas/adapter/http.py [hunk 1]
```diff
@@ -355,6 +355,15 @@ def request_body(cls, request: Request, expect_type: Type[T], stripped: bool) ->
...
 
+    @staticmethod
+    def decode_specific_asset_ids(encoded_ids: str):
+        # Decoding from base64url and transforming to JSON
+        decoded_json = base64.urlsafe_b64decode(encoded_ids + '===').decode('utf-8')
+        # Assuming the JSON represents a list of SpecificAssetId objects
+        specific_asset_ids = json.loads(decoded_json)
+        # Transform the list of dictionaries into SpecificAssetId instances
+        return [model.SpecificAssetId(**id_dict) for id_dict in specific_asset_ids]
+
```
Comment (jkhsjdhjs) [inline basyx/aas/adapter/http.py]: This method doesn't work (correctly) and should be removed. Instead, the function `base64urljson_list`, implemented in #22, should be used.

File: basyx/aas/adapter/http.py [hunk 2]
```diff
@@ -601,14 +610,23 @@ def handle_request(self, request: Request):
...
+        # Filtering by base64url encoded SpecificAssetIds if provided
         asset_ids = request.args.get("assetIds")
         if asset_ids is not None:
+            # Use the new method to decode and instantiate SpecificAssetIds
             spec_asset_ids = HTTPApiDecoder.base64urljson_list(asset_ids, model.SpecificAssetId, False, False)
-            # TODO: it's currently unclear how to filter with these SpecificAssetIds
-        return response_t(list(aas))
+            # Filter the AAS based on these SpecificAssetIds
+            aas = filter(lambda shell: all(
+                spec_id in [id_.identifier for id_ in shell.specific_asset_ids] for spec_id in spec_asset_ids), aas)
```
Comment (jkhsjdhjs) [inline basyx/aas/adapter/http.py]: `SpecificAssetId` doesn't have an `identifier` attribute, so this doesn't work. Furthermore, `AssetAdministrationShell` doesn't have a `specific_asset_ids` attribute. The `SpecificAssetIds` are located in the `AssetInformation`.

File: basyx/aas/adapter/http.py [hunk 3]
```diff
@@ -601,14 +610,23 @@ def handle_request(self, request: Request):
...
+
+        aas_list = list(aas)
```
Comment (jkhsjdhjs) [inline basyx/aas/adapter/http.py]: Why is this changed? This was perfectly fine before.

**Review by jkhsjdhjs — CHANGES_REQUESTED** [review]
This doesn't work, but the idea is correct. The changes required to make it work should be relatively simple.
Furthermore, I already rebased the branch to resolve conflicts that arose due to #22. Before you continue to work on this, you should do the following to remove the commit you have locally and pull the commit I rebased.
```
git reset --hard HEAD\^
git pull
```

**Comment by s-heppner** [comment]
After you did what @jkhsjdhjs  suggested, have a look at the StaticAnalysis Test on this page here: 

```
basyx/aas/adapter/http.py:617: error: "AssetInformation" has no attribute "specific_asset_ids"; maybe "specific_asset_id" or "_specific_asset_id"?  [attr-defined]
```

Other than that, the logic seems fine to me.

**Review by jkhsjdhjs — APPROVED** [review]
LGTM

**Pull request merged**